### PR TITLE
Timer0 ctc

### PR DESCRIPTION
**Pull request by paweo90**

### CODE REVIEW

File: src/avr/avr_m328p.c [hunk 1]
```diff
@@ -46,6 +46,9 @@
 
 #define NOT_CONNECTED		0xFFU
 
+static const uint8_t max_8bit = 0xFF;
+static const uint8_t bottom_8bit = 0x00;
```
Comment (dsalychev) [inline src/avr/avr_m328p.c]: This variable is not used anywhere. We may just remove it completely. But...

Why have you decided to create such variables? You use them as constants and a simple macro should be enough in this case. The most important difference here is a memory allocation. These two variables will _occupy memory_ in spite of the fact that they are used as constants, but macro will only be resolved during the compilation process and won't occupy any memory.

File: src/avr/avr_m328p.c [hunk 2]
```diff
@@ -259,6 +266,52 @@ static void timer0_normal(struct MSIM_AVR *mcu,
 	}
 }
 
+static void timer0_ctc(struct MSIM_AVR *mcu,
+                          uint32_t presc, uint32_t *ticks,
+                          uint8_t wgm0, uint8_t com0a, uint8_t com0b)
+{
+	uint8_t top = DM(OCR0A);
+
+	switch(wgm0)
```
Comment (dsalychev) [inline src/avr/avr_m328p.c]: Switch is here? Is it a typo?

File: src/avr/avr_m328p.c [hunk 3]
```diff
@@ -259,6 +266,52 @@ static void timer0_normal(struct MSIM_AVR *mcu,
...
+		if (DM(TCNT0) == max_8bit || DM(TCNT0) == top) {
+			/* Generate TOV flag always and only on 0xFF */
+			if(DM(TCNT0) == max_8bit) {
+				/* Set TimerCounter0 overflow flag  */
+				DM(TIFR0) |= DM(1<<TOV0);
+			}
+			/* Generate Interrupt on TOP value */
+			if (DM(TCNT0) == DM(OCR0A)) {
+				/* Set TC0 Output Compare A Flag */
+				DM(TIFR) |= (1<<OCF0A);
```
Comment (dsalychev) [inline src/avr/avr_m328p.c]: ATmega328 doesn't have a TIFR register. There are three actually: TIFR0, TIFR1 and TIFR2. You need the first one, TIFR0 to set OCF0A flag.

File: src/avr/avr_m328p.c [hunk 4]
```diff
@@ -259,6 +266,52 @@ static void timer0_normal(struct MSIM_AVR *mcu,
...
+				/* Manipulate on the OCR0A (PD6) pin  */
+				timer0_oc0_nonpwm(mcu, com0a, com0b, A_CHAN);
+			} else if (DM(TCNT0) == DM(OCR0B)) {
+				/* Set TC0 Output Compare B Flag */
+				DM(TIFR) |= (1<<OCF0B);
```
Comment (dsalychev) [inline src/avr/avr_m328p.c]: The same issue with TIFR register is here also.

**Review by dsalychev — CHANGES_REQUESTED** [review]

File: src/avr/avr_m328p.c (hunk 2 above)
Comment (paweo90) [reply src/avr/avr_m328p.c]: Copy paste problem, sorry

**Review by dsalychev — APPROVED** [review]

**Comment by dsalychev** [comment]
Done in `0.1.0134`: https://github.com/dsalychev/mcusim/commit/218d6da44c0de3e9718cee4506be33f9ea0c8774

**Comment by dsalychev** [comment]
@paweo90 Changes look reasonable. Good job :+1:

**Pull request closed without merging**